x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -2536,13 +2788,21 @@ private void verifyGetResponse(
                 assertThat(apiKey.getMetadata(), equalTo(metadata == null ? Map.of() : metadata));
             }
         }
-        Arrays.stream(response.getApiKeyInfos())
-            .forEach(
-                apiKeyInfo -> assertThat(
-                    apiKeyInfo.getRoleDescriptors(),
-                    containsInAnyOrder(expectedRoleDescriptors.toArray(RoleDescriptor[]::new))
-                )
-            );
+        Arrays.stream(response.getApiKeyInfos()).forEach(apiKeyInfo -> {
+            assertThat(apiKeyInfo.getRoleDescriptors(), containsInAnyOrder(expectedRoleDescriptors.toArray(RoleDescriptor[]::new)));
+            // No good way to test different limited-by role descriptors for different users
+            // It will be tested separately
+            if (user.length == 1) {
+                if (expectedLimitedRoleDescriptors == null) {
+                    assertThat(apiKeyInfo.getLimitedBy(), nullValue());
+                } else {
+                    assertThat(
+                        apiKeyInfo.getLimitedBy().roleDescriptorsList().get(0),
+                        containsInAnyOrder(expectedLimitedRoleDescriptors.toArray(RoleDescriptor[]::new))
+                    );
+                }
+            }
```

Comment:
Alternatively (this would be a bigger change - not necessary in this PR, and possibly not worth it) could pass a map of username to expected limitedBy RDs, e.g. something like:

```
Map.of(
  ES_TEST_ROOT_USER, List.of(ES_TEST_ROOT_ROLE_DESCRIPTOR), 
  "user_with_manage_api_key_role", List.of(...)
)
```

Reply:
I added the null assertion. 

> pass a map of username to expected limitedBy RDs

I considered that too but deciced not to do so because this multiple user assertion is only used in one place which does not really justify the cascading changes. There is definitely refactoring opportunities. But I'll leave it for now.

Reply:
On a second thought, I refactored this method to take lookup function to get expectedLimitedByRoleDescriptors for different users. This is because other changes such as randomisation (as suggested below) has made cascading changes no longer a valid concern. It is also good to get rid of the special handling inside this method.